services/login-settings/src/LoginSettingsProvider.tsx

```diff
@@ -0,0 +1,103 @@
+// import i18n from '@dhis2/d2-i18n'
+import { useDataQuery } from '@dhis2/app-service-data'
+import React, { useState } from 'react'
+import { LoginSettingsContext } from './LoginSettingsContext'
+import { LoginSettings, Locale } from './types'
+
+const querySettings = {
+    loginSettings: {
+        resource: 'auth/loginConfig',
+        params: ({ locale }: any) =>
+            locale
+                ? {
+                      paging: false,
+                      locale,
+                  }
+                : {
+                      paging: false,
+                  },
+    },
+}
+
+const queryLocales = {
+    localesUI: {
+        resource: 'locales/ui',
+    },
+}
+
+const localStorageLocaleKey = 'dhis2.locale.ui'
+
+const defaultLocales = [
+    { locale: 'en', name: 'English', displayName: 'English' },
+]
+
+const transformFetchedLocales = (fetchedLocales: any): Locale[] => {
+    if (Array.isArray(fetchedLocales)) {
+        return fetchedLocales.map((locale) => ({
+            name: locale?.name,
+            locale: locale?.locale,
+            displayName: locale?.displayName,
+        }))
+    }
+    return defaultLocales
+}
+
+const getLocale = (
+    selectedLocale: string | null,
+    systemUiLocale: string | undefined
+): string => {
+    if (selectedLocale !== null) {
+        return selectedLocale
+    }
+    if (systemUiLocale) {
+        return systemUiLocale
+    }
+    return 'en'
+}
+
+export const LoginSettingsProvider = ({
+    children,
+}: {
+    children: React.ReactNode
+}) => {
+    const { data, called, error, refetch } = useDataQuery(querySettings, {
+        variables: { locale: localStorage[localStorageLocaleKey] },
+    })
+
+    const {
+        data: dataLocales,
+        called: calledLocales,
+        error: errorLocales,
+    } = useDataQuery(queryLocales)
+
+    const [selectedLocale, setSelectedLocale] = useState(
+        window.localStorage.getItem(localStorageLocaleKey)
+    )
+
+    const refreshOnTranslation = ({ locale }: { locale: string }) => {
+        refetch({ locale })
+        setSelectedLocale(locale)
```

Comment:
I'm not totally sure how we will actually want to handle translations.

Presumably, we will have to make another request to retrieve system-specific translations (e.g. title) (hence this implementation) (but this depends on how https://dhis2.atlassian.net/browse/DHIS2-14681 is actually implemented)

Here, I am updating the uiLocale language before the refetch of these locales might be finished. This could produce some glitchiness because the ui translations would then load before the request finishes. We could hold off on using ui translations until request finishes, but then presumably we need UI feedback that translations are being retrieved (which might be kind of noisy given that most implementations don't seem to actually translate things like title and it's likely that only the ui language would translate anyway)